src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -380,24 +391,44 @@ static String minimize(String name) {
     @Extension
     public static class Deleter extends ItemListener {
 
-        private static /* almost final */ int CLEANUP_THREAD_LIMIT = SystemProperties.getInteger(Deleter.class.getName() + ".CLEANUP_THREAD_LIMIT", Integer.valueOf(0)).intValue();
+        private static final int CLEANUP_THREAD_LIMIT = SystemProperties.getInteger(Deleter.class.getName() + ".CLEANUP_THREAD_LIMIT", 0);
 
-        /** Semaphore for limiting number of scheduled {@link CleanupTask} */
-        private static Semaphore cleanupPool = new Semaphore(CLEANUP_THREAD_LIMIT, true);
+        private static final ExecutorService executorService = executorService();
 
         /** Number of {@link CleanupTask} which have been scheduled but not yet completed. */
         private static int runningTasks;
 
+        private static ExecutorService executorService() {
+            return CLEANUP_THREAD_LIMIT > 0
+                ? new ContextResettingExecutorService(
+                    new ImpersonatingExecutorService(
+                        new ErrorLoggingExecutorService(
+                            new ThreadPoolExecutor(0, CLEANUP_THREAD_LIMIT, 10L, TimeUnit.SECONDS, new SynchronousQueue<>(),
```

Comment:
New approach looks mostly good, but why `SynchronousQueue`? If I understand correctly, that will make it so that whenever there are more than `CLEANUP_THREAD_LIMIT` tasks, each submission will block until another task completes. I would use `LinkedBlockingQueue` instead so that when there is a burst of activity, we allow tasks to be queued up and do not block in `ItemListener.onDeleted`.

Reply:
Hmm, reading through https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ThreadPoolExecutor.html though it seems that using an unbounded queue would probably result in only ever starting a single thread, and worse, it seems like with `SynchronousQueue` we will actually _reject_ submitted tasks if we ever have more than `CLEANUP_THREAD_LIMIT` in flight at once.

Looking at Javadoc it does not seem possible to get the behavior we want with `ThreadPoolExecutor` alone. I guess the simplest thing would be to have the core pool size be `CLEANUP_THREAD_LIMIT`, leaving those threads basically permanently idle. Another approach could be to keep `SynchronousQueue`, but then have a separate unbounded queue that we add tasks to here, and then have a `PeriodicWork` or some other similar single-thread utility that repeatedly tries to submit tasks in the unbounded queue to the thread pool. Awkward.

Reply:
> Hmm, reading through https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ThreadPoolExecutor.html though it seems that using an unbounded queue would probably result in only ever starting a single thread, and worse, it seems like with SynchronousQueue we will actually reject submitted tasks if we ever have more than CLEANUP_THREAD_LIMIT in flight at once.

Correct. Tasks are being rejected..

> the core pool size be CLEANUP_THREAD_LIMIT, leaving those threads basically permanently idle

Seems to me that people impacted would put a value in the 1X / 1XX. And keeping 1XX threads alive might be overkill. But with a keepAlive time and [allowCoreThreadTimeOut](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ThreadPoolExecutor.html#allowCoreThreadTimeOut(boolean)) set to `true` this should have the behavior we want. Bringing threads up to the limit when needed, and then bring the pool down to `0` when no tasks are remaining and threads are becoming idle.

Reply:
Done. Couple of things that I tested:

* Startup size of the pool is `0`. With `CLEANUP_THREAD_LIMIT=100` on startup, I don't see any Deleter thread.
* Tasks enqueued when reaching the limit
* Threads are reuse. With `CLEANUP_THREAD_LIMIT=2` for the deletion of 4 branches across 4 nodes, I can see that the same thread (IDs) are being reused.

Reply:
Ah interesting, I wasn't familiar with `allowCoreThreadTimeOut`. That definitely looks useful here.